cylc/uiserver/scripts/gui.py

```diff
@@ -20,8 +20,165 @@
+        arg for arg in jp_server_opts if arg not in remove_these_args)
+    [sys.argv.remove(arg) for arg in remove_these_args]
+    if '--help' not in sys.argv:
+        # get existing jpserver-<pid>-open.html files
+        existing_guis = glob(os.path.join(INFO_FILES_DIR, "*open.html"))
+        if existing_guis and not new_gui:
+            gui_file = get_valid_existing_gui(existing_guis)
+            if gui_file:
+                print(
+                    "Opening with existing gui." +
+                    f" Use {CLI_OPT_NEW} option for a new gui.",
+                    file=sys.stderr
+                )
+                update_html_file(gui_file, workflow_id)
+                webbrowser.open(gui_file, autoraise=True)
+    return CylcUIServer.launch_instance(
+        jp_server_opts or None, workflow_id=workflow_id
+    )
+
+
+def get_arg_parser():
+    parser = argparse.ArgumentParser(add_help=False)
+    parser.add_argument(
+        CLI_OPT_NEW,
+        action='store_true',
+        default=False,
+        dest='new'
+    )
+    return parser
+
+
+def update_html_file(gui_file, workflow_id):
+    """ Update the html file to open at the correct workflow in the gui.
+    """
+    with open(gui_file, "r") as f:
+        file_content = f.read()
+    url_extract_regex = re.compile('url=(.*?)\"')
+    url_string = url_extract_regex.search(file_content)
+    if not url_string:
+        return
+    url = url_string.group(1)
+    split_url = url.split('/workflows/')
+    if not workflow_id:
+        # new url should point to dashboard
+        if len(split_url) == 1:
+            # no update required
+            return
+        else:
+            # previous url points to workflow page and needs updating
+            # replace with base url (including token)
+            replacement_url_string = split_url[0]
+    else:
+        if len(split_url) > 1:
+            old_workflow = split_url[1]
+            if workflow_id == old_workflow:
+                # same workflow page requested, no update needed
+                return
+            else:
+                replacement_url_string = url.replace(old_workflow, workflow_id)
+        else:
+            # current url points to dashboard, update to point to workflow
+            replacement_url_string = f"{url}/workflows/{workflow_id}"
+    update_url_string(gui_file, url, replacement_url_string)
+
+
+def update_url_string(gui_file: str, url: str, replacement_url_string: str):
+    """Updates the url string in the given gui file."""
+    file = Path(gui_file)
+    current_text = file.read_text()
+    updated_text = current_text.replace(url, replacement_url_string)
+    file.write_text(updated_text)
+
+
+def get_valid_existing_gui(gui_files: List[str]):
+    """Finds an existing gui to reopen"""
+    for file in gui_files:
+        pid_match = re.search(r"jpserver-(\d+)-open.html", file)
+        if pid_match is not None:
+            pid = pid_match[1]
+            if pid_exists(int(pid)):
```

Comment:
Spanner to throw into the works, the process might not be running on this server.

Does Jupyter Server have a mechanism for checking if a server is still running?

Reply:
I will have a look and report back, I had forgotten that case.

Reply:
So, Jupyter Server does have a command for checking if a server is running. I don't think this is of use to us though 😢 
The command is `jupyter server list`
- this checks for the existence of pid file (remember this doesn't get cleaned up if the server shuts down unexpectedly)
- it then does a `os.kill(pid, 0)` to check the process is running
- it defaults to process running in the case of `PermissionError(1, 'Operation not permitted')`, which I am meeting regularly as I have a load of stale files.
- the result is that it is claiming to have servers running when they are not.

If the process is running on a different server, maybe we can use the info in the pid file to pick the host out?
The info stored in the file is:
```
{
  "base_url": "/",
  "hostname": "localhost",
  "password": false,
  "pid": 11111,
  "port": 8888,
  "root_dir": "/net/home/../username",
  "secure": true,
  "sock": "",
  "token": "blah blah some big long token here",
  "url": "https://localhost:8888/",
  "version": "1.18.1"
} 
```